### PR TITLE
fix param_buffer peak memory may be OOM

### PR DESCRIPTION
**Pull request by wangxicoding**
## Problem description
As shown in Figure 2, using `torch.tensor(storage)` will increase the peak memory. `torch.tensor(storage)` may create a new tensor first, and then replace the storage of the new tensor with the incoming storage, resulting in a problem of high peak memory, and may cause OOM. (For example, if we have 16GB param and 32GB grad, and when obtaining the param_buffer 32GB storage is needed, 16+32+32 = 80GB, it may cause OOM)
## Solution
As shown in Figure 1, we can use `tensor.view(dtype)` to share tensor, there will be no problem with high peak memory. In addition, as shown in Figure 3, `.view()` does not ignore views/slices, so there is no need to calculate offset.

ubFd36nrJe

### CODE REVIEW

**Comment by deepakn94** [comment]
Did you try running this with `--use-distributed-optimizer --overlap-grad-reduce`?

**Comment by wangxicoding** [comment]
> Did you try running this with `--use-distributed-optimizer --overlap-grad-reduce`?

I have tested `len(grad_buffer.buckets) > 1`, as expected.
Megatron doesn't have CI on GitHub, maybe you can use the internal CI to test it again.

**Comment by yiakwy-xpu-ml-framework-team** [comment]
Hi @wangxicoding there are codes in megatron where you can make variable inplace version (loss layer for example), which could save you around activation 20 GB. Besides tensor view, you can re-use some tensor's memory in-place.

Wish this  helps you.

**Comment by deepakn94** [comment]
Thanks a lot for this patch, @wangxicoding ! We merged it internally, and it should be pushed to Github soon!

**Comment by wangxicoding** [comment]
> Thanks a lot for this patch, @wangxicoding ! We merged it internally, and it should be pushed to Github soon!

ok, thanks
